interfaces/apparmor/spec.go

```diff
@@ -160,6 +160,31 @@ func (spec *Specification) AddSnapLayout(si *snap.Info) {
 	}
 }
 
+// AddParallelInstanceMapping adds AppArmor snippets allowing remapping of snap
+// directories for parallel installed snaps
+//
+// Specifically snap-update-ns will apply the following bind mounts
+// - /snap/foo_bar -> /snap/foo
+// - /var/snap/foo_bar -> /var/snap/foo
+// - /home/joe/snap/foo_bar -> /home/joe/snap/foo
+func (spec *Specification) AddParallelInstanceMapping(si *snap.Info) {
+	if si.InstanceKey == "" {
+		return
+	}
+	var buf bytes.Buffer
+
+	// /snap/foo_bar -> /snap/foo
+	fmt.Fprintf(&buf, "  # Allow parallel instance snap mount namespace adjustments\n")
+	fmt.Fprintf(&buf, "  mount options=(rw rbind) /snap/%s/ -> /snap/%s/,\n", si.InstanceName(), si.SnapName())
```

Comment:
I think we should be OK with just a read-only mount but I may be missing some context. I worry that perhaps in a case where $SNAP is used as a tmpfs-based collector of plugins it may be really needed to use `rw` here. In either case can you please add a comment that explains why we are using writable bind mounts here. It will be useful for us later.

Reply:
Agree with zyga, can we please make this read-only for the time being?  I can't see why we'd write there. For the tmpfs case zyga mentioned, it's a different filesystem which is mounted on top of it.

Reply:
To complicate the matter we must do `rbind` and then `ro` each _created_ mount point manually because kernel sucks and doesn't have nice APIs for this. Specifically we must do `rbind`, enumerate the mount points that were created and then do `remount,bind,ro` on each one that we want to have read-only.

Reply:
Maciej and me had a call about this issue and we managed to simplify instances even more to the point where we don't need to ensure read-only bind mounts (which need to be done by hand).

The idea is as follows:

1) We rbind mount the instance specific directory (for $SNAP, $SNAP_DATA and $SNAP_USER_DATA) as the first thing in the mount profile (or per user mount profile). This bind mount uses our secure code path that is immune to symlink attacks.
2) We switch all other mount backend operations to be unaware of instances. They will perform their changes in the per-snap (instance) and per-user (instance) mount namespace in the usual location which because of 1) contains instance-specific code and data

This matches the recent simplification which removed various instance specific variables. Once achieved the only way to know you are an instance is to look at the SNAP_INSTANCE_KEY environment variable. Everything else will be exactly identical to non-instance behaviour.

Because 1) is executed early (before layout and regular mount interface operations) we have a guarantee that we only replicate instance revisions (mount points of each revision of an instance) and thus don't need to implement the extra code that would enumerate those and switch each one to a read-only bind mount.

Subsequent operations that connect content interface data to $SNAP or $SNAP_DATA will be no different than non-instance counterparts.